src/functionalTest/java/uk/gov/hmcts/reform/bulkscanprocessor/controllers/TestHelper.java

```diff
@@ -140,24 +146,43 @@ public String getRandomFilename(String suffix) {
     ) throws IOException {
         ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
         try (ZipOutputStream zos = new ZipOutputStream(outputStream)) {
+            Map<String, String> tokens = new ConcurrentHashMap<>();
+
             for (String file : files) {
-                zos.putNextEntry(new ZipEntry(file));
+                String dcn = generateDcnNumber();
+                tokens.put("##" + file.replace(".pdf", "") + "##", dcn);
+                zos.putNextEntry(new ZipEntry(dcn + ".pdf"));
                 zos.write(Resources.toByteArray(Resources.getResource(file)));
                 zos.closeEntry();
             }
 
             if (metadataFile != null) {
                 String metadataTemplate =
                     Resources.toString(Resources.getResource(metadataFile), StandardCharsets.UTF_8);
-                String metadata = metadataTemplate.replace("$$zip_file_name$$", zipFilename);
                 zos.putNextEntry(new ZipEntry("metadata.json"));
-                zos.write(metadata.getBytes());
+                tokens.put("##zip_file_name##", zipFilename);
+
+                Pattern pattern = Pattern.compile("(" + String.join("|", tokens.keySet()) + ")");
+                Matcher matcher = pattern.matcher(metadataTemplate);
+                StringBuffer sb = new StringBuffer();
+
+                while (matcher.find()) {
+                    matcher.appendReplacement(sb, tokens.get(matcher.group(1)));
+                }
+
+                matcher.appendTail(sb);
+
+                zos.write(sb.toString().getBytes());
```

Comment:
The part where things are replaced in metadata file is very hard to follow. Maybe it could be simplified or extracted to a separate method?

Reply:
Could be as simply iterate over `token` and do `replaceAll`.
No performance to gain imho if that's what you were after here originally.

Reply:
That's what I was thinking, too. It will only be negligibly slower, i.e. no difference to be noticed. And we're talking about tests. The benefit will be simplicity.

Reply:
`replaceAll` replaces any regex occurrence with a single (same) string. there are metafiles with 2 different scannable items

Reply:
I was thinking about something more or less like this:
```
        tokens
            .entrySet()
            .stream()
            .reduce(
                metadataTemplate,
               (data, token) -> data.replaceAll(token.getKey(), token.getValue())
            );
```

Whatever the solution, it would be really good to have all the field value replacement in a separate method.

Reply:
I was thinking about something more or less like this:

        tokens
            .keySet()
            .stream()
            .reduce(
                metadataTemplate,
               (data, token) -> data.replaceAll(token, tokens.get(token))
            );
Whatever the solution, it would be really good to have all the field value replacement in a separate method.